src/main/java/io/gravitee/policy/assignattributes/AssignAttributesPolicy.java

```
@@ -15,158 +15,97 @@
-
-                    super.end();
-                }
-            };
-        }
+    @Override
+    public Completable onRequest(HttpExecutionContext ctx) {
+        return assign(ctx);
+    }
 
-        return null;
+    @Override
+    public Completable onResponse(HttpExecutionContext ctx) {
+        return assign(ctx);
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        if (assignVariablePolicyConfiguration.getScope() == null || assignVariablePolicyConfiguration.getScope() == PolicyScope.REQUEST) {
-            // assign
-            assign(executionContext);
-        }
+    @Override
+    public Completable onMessageRequest(MessageExecutionContext ctx) {
+        return ctx.request().onMessage(message -> assign(ctx, message));
+    }
 
-        // continue chaining
-        policyChain.doNext(request, response);
+    @Override
+    public Completable onMessageResponse(MessageExecutionContext ctx) {
+        return ctx.response().onMessage(message -> assign(ctx, message));
     }
 
-    @OnResponse
-    public void onResponse(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        if (assignVariablePolicyConfiguration.getScope() != null && assignVariablePolicyConfiguration.getScope() == PolicyScope.RESPONSE) {
-            // assign
-            assign(executionContext);
-        }
+    private Completable assign(HttpExecutionContext executionContext) {
+        return filterAttributes()
+            .flatMapMaybe(attribute ->
+                executionContext
+                    .getTemplateEngine()
+                    .eval(attribute.getValue(), Object.class)
+                    .doOnSuccess(extValue -> executionContext.setAttribute(attribute.getName(), extValue))
+            )
+            .ignoreElements()
+            .onErrorResumeWith(
+                executionContext.interruptWith(
```

That sounds great to me.
As discussed, I prefer having an explicit error than just a log (inaccessible from the api publisher most of the time). So it means that there is a real BREAKING CHANGE now 🙂
@guillaumelamirand @brasseld can we have your feeling about breaking the execution when there is trouble with assigning attributes (ex: EL expression fails) instead of just logging? Are you comfortable with this BC?

---

If I remember, for conditions we distinguished two types of errors ?
- the EL is invalid because of bad syntax, so it should break
- the EL try to access a null value (the case of an absent header for stuff like `request.headers[X-Toto][0] == 'totovalue'`), in this case it's another exception, and it is just skipped to not break.

Maybe we could apply the same kind of logic here:
- in case of invalid syntax, we break
- in case of evaluation problem, the attribute is just not set 

wdyt ?

---

I looked at the gravitee-expression-language repo and I did found two exceptions like you said, but the one for the null value (ELNullEvaluationException) is not used. We only throw the other one ExpressionEvaluationException I think :/